### ISSUE TITLE
Looking forward to the code

### ISSUE THREAD

**Issue by Wuziyi616** [subhc/unsup-parts#1]
Thanks for the great work and your effort in releasing the code! Is there an expected date that your code will be released? Thanks in advance.

**Comment by subhc** [comment]
Thank you for your interest! We are hoping to release it by the NeurIPS conference.

**Comment by Wuziyi616** [comment]
Thanks for the reply!

**Comment by Wuziyi616** [comment]
Just a quick question. In eq(1) the feature loss, do you use the hard mask (taking after argmax) as M to construct the average part code z_k(I), or do you use the soft mask (after softmax) to compute the average vector as an expectation over pixel features?

**Comment by subhc** [comment]
We use the softmasks as with hard masks it would be hard to train.

```
 _, _, pred_low, _ = self.model(images)
pred = self.interp(pred_low)

zoo_feat = self.get_zoo_feat(images_vgg)
zoo_feat = zoo_feat * mask

loss_feat = 0
for pred_sm_, zoo_feat_, mask_ in zip(pred_softmax * mask, zoo_feat, mask): # to fit in GPU
    loss_feat += loss.consistency_loss(pred_sm_.unsqueeze(0), zoo_feat_.unsqueeze(0), mask_.unsqueeze(0))
loss_feat_value += self.args.lambda_feat * loss_feat.data.cpu().numpy()

loss_rgb = loss.consistency_loss(pred_softmax * mask, images_rec, mask)
loss_rgb_value += self.args.lambda_rgb * loss_rgb.data.cpu().numpy()

# contrastive_loss
Z = torch.einsum('brhw, bchw -> brc', pred_softmax * mask, zoo_feat)
Z /= einops.reduce(pred_softmax * mask, 'b r h w -> b r 1', 'sum') + 1e-7
loss_contrastive = loss.contrastive_loss(Z, self.args.temperature)
loss_contrastive_value += self.args.lambda_contrastive * loss_contrastive.data.cpu().numpy()

```

```
def contrastive_loss(feats, t=0.07):
    feats = F.normalize(feats, dim=2)  # B x K x C
    scores = torch.einsum('aid, bjd -> abij', feats, feats)
    scores = einops.rearrange(scores, 'a b i j -> (a i) (b j)')

    # positive logits: Nx1
    pos_idx = einops.repeat(torch.eye(feats.size(1), dtype=torch.int, device=feats.device), 'i j -> (a i) (b j)', a=feats.size(0), b=feats.size(0))
    pos_idx.fill_diagonal_(0)
    l_pos = torch.gather(scores, 1, pos_idx.nonzero()[:, 1].view(scores.size(0), -1))
    rand_idx = torch.randint(1, l_pos.size(1), (l_pos.size(0), 1), device=feats.device)
    l_pos = torch.gather(l_pos, 1, rand_idx)

    # negative logits: Nx(K-1)
    neg_idx = einops.repeat(1-torch.eye(feats.size(1), dtype=torch.int, device=feats.device), 'i j -> (a i) (b j)', a=feats.size(0), b=feats.size(0))
    l_neg = torch.gather(scores, 1, neg_idx.nonzero()[:, 1].view(scores.size(0), -1))

    # logits: NxK
    logits = torch.cat([l_pos, l_neg], dim=1)

    # apply temperature
    logits /= t

    # labels: positive is at 0th index
    labels = torch.zeros(logits.shape[0], dtype=torch.long, device=scores.device)
    return F.cross_entropy(logits, labels)

def consistency_loss(masks, image, fg):
    # masks: N x R x H x W
    # image: N x 3 x H x W
    weighted_regions = masks.unsqueeze(2) * image.unsqueeze(1)  # N x R x 3 x H x W
    mask_sum = masks.sum(3).sum(2, keepdim=True)  # N x R x 1
    means = weighted_regions.sum(4).sum(3) / (mask_sum + 1e-5)  # N x R x 3
    diff_sq = (image.unsqueeze(1) - means.unsqueeze(3).unsqueeze(4))**2  # N x R x 3 x H x W
    loss = (diff_sq * masks.unsqueeze(2)*fg.unsqueeze(2)).sum(4).sum(3)  # N x R x 3
    loss /= (fg.unsqueeze(2).sum(4).sum(3) + 1e-5)  # N x R x 3
    return loss.sum(2).sum(1).mean()
```
We build upon SCOPS' code so would be good to check that out. I have to make sure if this is the version I used at the end but you get the idea.

**Comment by Wuziyi616** [comment]
Yea that's very helpful. Thanks!

**Comment by subhc** [comment]
I have updated the repo to include the code, thanks.
